libmamba/tests/src/util/test_iterator.cpp

```diff
@@ -0,0 +1,198 @@
+
+        return std::make_tuple(input_sequence, result_sequence, input_map, result_map);
+
+    }
+}
+
+TEST_SUITE("filter_iterator")
+{
+    TEST_CASE("forward_iterator_api")
+    {
+        auto [input_seq, res_seq, input_map, res_map] = get_test_data();
+        {
+            auto iter = filter_begin(greater_than_10, input_seq);
+            auto iter_end = filter_end(greater_than_10, input_seq);
+
+            auto citer = cfilter_begin(greater_than_10, input_seq);
+            auto citer_end = cfilter_end(greater_than_10, input_seq);
+
+            auto res_iter = res_seq.begin();
+
+            auto iter2 = iter, iter3 = iter;
+
+            for (size_t i = 0; i < res_seq.size(); ++i)
+            {
+                CHECK_EQ(*iter, *res_iter);
+                CHECK_EQ(*iter, *citer);
+                ++iter, ++citer, ++res_iter;
+            }
+            CHECK_EQ(iter, iter_end);
+            CHECK_EQ(citer, citer_end);
+            CHECK(iter == iter_end);
+
+            CHECK_EQ(iter2.operator->(), &input_seq[1]);
+            CHECK(iter2++ != ++iter3);
+        }
+        {
+            auto iter = filter_begin(second_greater_than_10, input_map);
+            auto iter_end = filter_end(second_greater_than_10, input_map);
+
+            auto citer = cfilter_begin(second_greater_than_10, input_map);
+            auto citer_end = cfilter_end(second_greater_than_10, input_map);
+
+            auto res_iter = res_map.begin();
+
+            auto iter2 = iter, iter3 = iter;
+
+            for (size_t i = 0; i < res_map.size(); ++i)
+            {
+                CHECK_EQ(*iter, *res_iter);
+                CHECK_EQ(*iter, *citer);
+                ++iter, ++citer, ++res_iter;
+            }
+            CHECK_EQ(iter, iter_end);
+            CHECK_EQ(citer, citer_end);
+            CHECK(iter == iter_end);
+            
+            CHECK_EQ(iter2->second, res_map.begin()->second);
+            CHECK(iter2++ != ++iter3);
+        }
+    }
+
+    TEST_CASE("bidirectional_iterator_api")
+    {
+        auto [input_seq, res_seq, input_map, res_map] = get_test_data();
+        {
+            auto iter = filter_begin(greater_than_10, input_seq);
+            auto iter_end = filter_end(greater_than_10, input_seq);
+
+            auto citer = cfilter_begin(greater_than_10, input_seq);
+            auto citer_end = cfilter_end(greater_than_10, input_seq);
+
+            auto res_iter = res_seq.end();
+            
+            auto iter2 = iter_end, iter3 = iter_end;
+
+            for (size_t i = 0; i < res_seq.size(); ++i)
+            {
+                --iter_end, --citer_end, --res_iter;
+                CHECK_EQ(*iter_end, *res_iter);
+                CHECK_EQ(*iter_end, *citer_end);
+            }
+            CHECK_EQ(iter, iter_end);
+            CHECK_EQ(citer, citer_end);
+            --iter_end, --citer_end;
+            CHECK_EQ(iter_end.operator->(), (--input_seq.begin()).operator->());
+
+            CHECK(iter2-- != --iter3);
```

Comment:
Is order of invocation important here? The order of invocation have changed in C++17 if my memory is correct, it is now guaranteed that the left side is executed before the right side. To clarify: this was not guaranteed in C++14 and before. 
It's also cryptic to read, so I would prefer with the 2 operations separated, as we're in tests it's ok if we name temporaries for clarity.

Reply:
The order order of invocation is not important, the idea is to test that `operator--` and `operator--(int)` give different results.